src/test/mir-opt/validate_3.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-tidy-linelength
+// compile-flags: -Z verbose -Z mir-emit-validate
+
+struct Test {
+    x: i32
+}
+
+fn foo(_x: &i32) {}
+
+fn main() {
+    let t = Test { x: 0 };
+    let t = &t;
+    foo(&t.x);
```

Comment:
Reacquire happens at the EndRegion, so code generation is no different. I left the relevant EndRegion in the tests for this reason.

I also hope that sooner rather than later after this is merged, the miri test suite will start testing these things, and with miri merged in rustc, maybe we can run the miri test suite with this "run cargo test on external projects" infrastructure we got.